### PR TITLE
[PyTorch] Fix waste in unfold()

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55339 [PyTorch] Fix waste in unfold()**
* #55336 [PyTorch] Cache self.size(dim) in TensorShape functions

Use DimVector. Avoid calling size()/stride() when we know argument is in bounds.

Differential Revision: [D27577647](https://our.internmc.facebook.com/intern/diff/D27577647/)

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -2133,19 +2133,22 @@ Tensor alias(const Tensor& self) {
 Tensor unfold(const Tensor& self, int64_t dimension, int64_t size, int64_t step) {
   // some special handling to deal with allow dimension == 0 when self.dim() == 0
   dimension = at::maybe_wrap_dim(dimension, self.dim(), /*wrap_scalar=*/true);
-  int64_t max_size = self.dim() == 0 ? 1 : self.size(dimension);
+
+  const auto sizes = self.sizes();
+  const auto strides = self.strides();
+  int64_t max_size = self.dim() == 0 ? 1 : sizes[0];
```
Comment (ngimel) [inline aten/src/ATen/native/TensorShape.cpp:2139]: it used to be `sizes[dimension]`, why the change?

Comment (ngimel) [reply aten/src/ATen/native/TensorShape.cpp:2139]: Also, `fold` accepts negative dimension, so you have to do `maybe_wrap_dim`

Comment (swolchok) [reply aten/src/ATen/native/TensorShape.cpp]: it's wrapped on line 2134 above (the first line of the function), isn't it?

Comment (ngimel) [reply aten/src/ATen/native/TensorShape.cpp]: Oh, my bad, I'm blind.
